File: extractor/src/Quest/DataStructures/Quest.ts
import { Expose, Type } from "class-transformer";
import { ArrayEqual } from "../../CompareUtils/ArrayEqual.js";
import { RecordObject } from "../../Database/RecordObject.js";
import { QuestRewardItem } from "./QuestRewardItem.js";
import { SqlField } from "../../SqlConverter/Decorators/SqlField.js";

/**
 * Represents a Quest in the tool.
 */
export class Quest implements RecordObject {
	/**
	 * The File Version that originated this object
	 */
	@Expose()
	@SqlField()
	public _FileVersion: number = -1;

	/**
	 * Quest ID
	 */
	@Expose()
	@SqlField()
	public Id: number = 0;

	/**
	 * Quest title ("yellow" title in side UI, or entry name in quest UI list)
	 */
	@Expose()
	@SqlField()
	public Title: string = "";

	/**
	 * Quest long description inside quest UI list
	 */
	@Expose()
	@SqlField({
		transform: (value: string[]) => value.join('\n'),
		outType: () => String,
	})
	public Description: string[] = [];

	/**
	 * Short, one-line, mission description. Shown in side UI.
	 */
	@Expose()
	@SqlField()
	public Summary: string = "";

	/**
	 * Always "QUE_NOIMAGE" or empty string (""), used to be something when viewing the quest,
	 * but was not really used officially
	 */
	@Expose()
	@SqlField()
	public OldImage: string = "";

	/**
	 * Icon shown in the quest list.
	 * In the old UI (before 2018-03-21), that was always SG_FEEL,
	 * In the new UI, this uses a few images that represents "type of quest"
	 */
	@Expose()
	@SqlField()
	public IconName: string = "";

	/**
	 * NPC Sprite
	 */
	@Expose()
	@SqlField()
	public NpcSpr: string = "";

	/**
	 * NPC Map
	 */
	@Expose()
	@SqlField()
	public NpcNavi: string = "";

	/**
	 * NPC X
	 */
	@Expose()
	@SqlField()
	public NpcPosX: number = -1;

	/**
	 * NPC Y
	 */
	@Expose()
	@SqlField()
	public NpcPosY: number = -1;

	/**
	 * Rewarded Base EXP
	 */
	@Expose()
	@SqlField()
	public RewardEXP: string = "";

	/**
	 * Rewarded Job EXP
	 */
	@Expose()
	@SqlField()
	public RewardJEXP: string = "";

	/**
	 * Rewarded items
	 */
	@Expose()
	@Type(() => QuestRewardItem)
	@SqlField({
		outType: () => String,
		transform: (value: QuestRewardItem[]) => JSON.stringify(value),
	})
	public RewardItemList: QuestRewardItem[] = [];

	/**
	 * Is it a Cooldown quest (?)
	 */
	@Expose()
	@SqlField()
	public CoolTimeQuest: number = -1;

	public getId(): string {
		return this.Id.toString();
	}

	public getFileVersion(): number {
		return this._FileVersion;
	}

	public equals(other: Quest): boolean {
		// _FileVersion is not checked, if 2 versions exists but the record is the same, we don't care.
		return (
			other.Title === this.Title
			&& ArrayEqual.isEqual(other.Description, this.Description)
			&& other.Summary === this.Summary
			&& other.IconName === this.IconName
			&& other.NpcSpr === this.NpcSpr
			&& other.NpcNavi === this.NpcNavi
			&& other.NpcPosX === this.NpcPosX
			&& other.NpcPosY === this.NpcPosY
			&& other.RewardEXP === this.RewardEXP
			&& other.RewardJEXP === this.RewardJEXP
			&& ArrayEqual.isEqual(this.RewardItemList, other.RewardItemList, (a, b) => a.equals(b))
		);
	}
}

File: extractor-go/internal/ro/rostructs/Quest.go
package rostructs

import (
	"strings"

	"github.com/guilherme-gm/ro-vis/extractor/internal/database/dao"
	"github.com/guilherme-gm/ro-vis/extractor/internal/domain"
)

type QuestV1 struct {
	QuestId     int
	Title       string
	Icon        string
	Image       string
	Description string
	Summary     string
}

func (q *QuestV1) ToDomain() domain.Quest {
	return domain.Quest{
		QuestID:     int32(q.QuestId),
		FileVersion: 1,
		Title:       dao.ToNullableString(q.Title),
		IconName:    dao.ToNullableString(q.Icon),
		OldImage:    dao.ToNullableString(q.Image),
		Summary:     dao.ToNullableString(q.Summary),
		Description: dao.ToNullableString(q.Description),
	}
}

type QuestV3RewardItem struct {
	ItemID  int
	ItemNum int
}

type QuestV3 struct {
	QuestId        int `lua:"@index"`
	Title          string
	Description    []string
	Summary        string
	IconName       string
	NpcSpr         string
	NpcNavi        string
	NpcPosX        int
	NpcPosY        int
	RewardEXP      string
	RewardJEXP     string
	RewardItemList []QuestV3RewardItem
}

func (q *QuestV3) ToDomain() domain.Quest {
	return domain.Quest{
		QuestID:     int32(q.QuestId),
		FileVersion: 3,
		Title:       dao.ToNullableString(q.Title),
		Description: dao.ToNullableString(strings.Join(q.Description, "\n")),
		Summary:     dao.ToNullableString(q.Summary),
		IconName:    dao.ToNullableString(q.IconName),
		NpcSpr:      dao.ToNullableString(q.NpcSpr),
		NpcNavi:     dao.ToNullableString(q.NpcNavi),
		NpcPosX:     dao.ToNullableInt32(int32(q.NpcPosX)),
		NpcPosY:     dao.ToNullableInt32(int32(q.NpcPosY)),
		RewardEXP:   dao.ToNullableString(q.RewardEXP),
		RewardJEXP:  dao.ToNullableString(q.RewardJEXP),
		// TODO:
		// RewardItemList: dao.ToNullString(q.RewardItemList),
	}
}

type QuestV4 struct {
	QuestId        int `lua:"@index"`
	Title          string
	Description    []string
	Summary        string
	IconName       string
	NpcSpr         string
	NpcNavi        string
	NpcPosX        int
	NpcPosY        int
	RewardEXP      string
	RewardJEXP     string
	RewardItemList []QuestV4RewardItem
	// New in V4
	CoolTimeQuest int
}

type QuestV4RewardItem struct {
	ItemID  int
	ItemNum int
}

func (q *QuestV4) ToDomain() domain.Quest {
	return domain.Quest{
		QuestID:     int32(q.QuestId),
		FileVersion: 3,
		Title:       dao.ToNullableString(q.Title),
		Description: dao.ToNullableString(strings.Join(q.Description, "\n")),
		Summary:     dao.ToNullableString(q.Summary),
		IconName:    dao.ToNullableString(q.IconName),
		NpcSpr:      dao.ToNullableString(q.NpcSpr),
		NpcNavi:     dao.ToNullableString(q.NpcNavi),
		NpcPosX:     dao.ToNullableInt32(int32(q.NpcPosX)),
		NpcPosY:     dao.ToNullableInt32(int32(q.NpcPosY)),
		RewardEXP:   dao.ToNullableString(q.RewardEXP),
		RewardJEXP:  dao.ToNullableString(q.RewardJEXP),
		// TODO:
		// RewardItemList: dao.ToNullString(q.RewardItemList),
	}
}
